File: ios/TextileBridge.m
#import "TextileBridge.h"

#if __has_include(<React/RCTBridge.h>)
#import <React/RCTBridge.h>
#elif __has_include(“RCTBridge.h”)
#import “RCTBridge.h”
#else
#import “React/RCTBridge.h”
#endif

#import <Textile/TextileApi.h>
#import "TextileEvents.h"
#import "utils.h"

#define SYSTEM_VERSION_LESS_THAN(v) ([[[UIDevice currentDevice] systemVersion] compare:v options:NSNumericSearch] == NSOrderedAscending)

@implementation TextileBridge

@synthesize bridge = _bridge;

RCT_EXPORT_MODULE();

- (void)invalidate {
  dispatch_semaphore_t sema = dispatch_semaphore_create(0);
  dispatch_async([self methodQueue], ^{
    [Textile.instance destroyWithCompletion:^(BOOL success, NSError *error){
      dispatch_semaphore_signal(sema);
    }];
  });
  dispatch_semaphore_wait(sema, DISPATCH_TIME_FOREVER);
}

- (dispatch_queue_t)methodQueue {
  return dispatch_queue_create("io.textile.TextileNodeQueue", DISPATCH_QUEUE_SERIAL);
}

RCT_EXPORT_METHOD(newWallet:(NSInteger)wordCount resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
  NSError *error;
  NSString *phrase = [Textile newWallet:wordCount error:&error];
  fulfillWithResult(phrase, error, resolve, reject);
}

RCT_EXPORT_METHOD(walletAccountAt:(NSString *)phrase index:(NSInteger)index password:(NSString *)password resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
  NSError *error;
  MobileWalletAccount *account = [Textile walletAccountAt:phrase index:index password:password error:&error];
  fulfillWithResult([account.data base64EncodedStringWithOptions:0], error, resolve, reject);
}

RCT_EXPORT_METHOD(isInitialized:(NSString *)repoPath resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
  BOOL initialized = [Textile isInitialized:repoPath];
  fulfillWithResult(@(initialized), nil, resolve, reject);
}

RCT_EXPORT_METHOD(initialize:(NSString *)repoPath seed:(NSString *)seed debug:(BOOL)debug logToDisk:(BOOL)logToDisk resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
  NSError *error;
  [Textile initialize:repoPath seed:seed debug:debug logToDisk:logToDisk error:&error];
  fulfillWithResult(nil, error, resolve, reject);
}

RCT_EXPORT_METHOD(initializeCreatingNewWalletAndAccount:(NSString *)repoPath debug:(BOOL)debug logToDisk:(BOOL)logToDisk resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
  NSError *error;
  NSString *phrase = [Textile initializeCreatingNewWalletAndAccount:repoPath debug:debug logToDisk:logToDisk error:&error];
  fulfillWithResult(phrase, error, resolve, reject);
}

RCT_EXPORT_METHOD(launch:(NSString *)repoPath debug:(BOOL)debug resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
  NSError *error;
  BOOL success = [Textile launch:repoPath debug:debug error:&error];
  if (success) {
    TextileEvents *eventHandler = self.bridge.textileEvents;
    Textile.instance.delegate = eventHandler;
  }
  fulfillWithResult(nil, error, resolve, reject);
}

RCT_EXPORT_METHOD(version:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
  resolve(Textile.instance.version);
}

RCT_EXPORT_METHOD(gitSummary:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
  resolve(Textile.instance.gitSummary);
}

RCT_EXPORT_METHOD(summary:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
  NSError *error;
  Summary *summary = [Textile.instance summary:&error];
  fulfillWithResult([summary.data base64EncodedStringWithOptions:0], error, resolve, reject);
}

@end

@implementation RCTBridge (TextileBridge)

- (TextileBridge *)textileBridge {
  return [self moduleForClass:[TextileBridge class]];
}

@end

File: android/src/main/java/io/textile/rnmobile/TextileBridge.java
package io.textile.rnmobile;

import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.modules.core.DeviceEventManagerModule;

import java.util.concurrent.Executor;
import java.util.concurrent.Executors;

import io.textile.pb.Mobile;
import io.textile.pb.View;
import io.textile.textile.Textile;

public class TextileBridge extends ReactContextBaseJavaModule {

    private Executor executor = Executors.newSingleThreadExecutor();
    private ReactApplicationContext context;

    public TextileBridge(ReactApplicationContext reactContext) {
        super(reactContext);
        this.context = reactContext;
    }

    @Override
    public String getName() {
        return "TextileBridge";
    }

    @ReactMethod
    public void newWallet(final long wordCount, final Promise promise) {
        executor.execute(new Runnable() {
            @Override
            public void run() {
                try {
                    promise.resolve(Textile.newWallet(wordCount));
                } catch (final Exception e) {
                    promise.reject("newWallet", e);
                }
            }
        });
    }

    @ReactMethod
    public void walletAccountAt(final String phrase, final long index, final String password, final Promise promise) {
        executor.execute(new Runnable() {
            @Override
            public void run() {
                try {
                    Mobile.MobileWalletAccount account = Textile.walletAccountAt(phrase, index, password);
                    promise.resolve(Util.encode(account.toByteArray()));
                } catch (final Exception e) {
                    promise.reject("walletAccountAt", e);
                }
            }
        });
    }

    @ReactMethod
    public void isInitialized(final String repoPath, final Promise promise) {
        executor.execute(new Runnable() {
            @Override
            public void run() {
                try {
                    promise.resolve(Textile.isInitialized(repoPath));
                } catch (final Exception e) {
                    promise.reject("isInitialized", e);
                }
            }
        });
    }

    @ReactMethod
    public void initialize(final String repoPath, final String seed, final Boolean debug, final Boolean logToDisk, final Promise promise) {
        executor.execute(new Runnable() {
            @Override
            public void run() {
                try {
                    Textile.initialize(repoPath, seed, debug, logToDisk);
                    promise.resolve(null);
                } catch (final Exception e) {
                    promise.reject("initialize", e);
                }
            }
        });
    }

    @ReactMethod
    public void initializeCreatingNewWalletAndAccount(final String repoPath, final Boolean debug, final Boolean logToDisk, final Promise promise) {
        executor.execute(new Runnable() {
            @Override
            public void run() {
                try {
                    promise.resolve(Textile.initializeCreatingNewWalletAndAccount(repoPath, debug, logToDisk));
                } catch (final Exception e) {
                    promise.reject("initializeCreatingNewWalletAndAccount", e);
                }
            }
        });
    }

    @ReactMethod
    public void launch(final String repoPath, final Boolean debug, final Promise promise) {
        executor.execute(new Runnable() {
            @Override
            public void run() {
                try {
                    Textile.launch(context, repoPath, debug);
                    DeviceEventManagerModule.RCTDeviceEventEmitter emitter =
                            context.getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class);
                    Textile.instance().addEventListener(new TextileEvents(emitter));
                    promise.resolve(null);
                } catch (final Exception e) {
                    promise.reject("launch", e);
                }
            }
        });
    }

    @ReactMethod
    public void version(final Promise promise) {
        executor.execute(new Runnable() {
            @Override
            public void run() {
                try {
                    promise.resolve(Textile.instance().version());
                }
                catch (final Exception e) {
                    promise.reject("version", e);
                }
            }
        });
    }

    @ReactMethod
    public void gitSummary(final Promise promise) {
        executor.execute(new Runnable() {
            @Override
            public void run() {
                try {
                    promise.resolve(Textile.instance().gitSummary());
                }
                catch (final Exception e) {
                    promise.reject("gitSummary", e);
                }
            }
        });
    }

    @ReactMethod
    public void summary(final Promise promise) {
        executor.execute(new Runnable() {
            @Override
            public void run() {
                try {
                    final View.Summary summary = Textile.instance().summary();
                    promise.resolve(Util.encode(summary.toByteArray()));
                }
                catch (final Exception e) {
                    promise.reject("summary", e);
                }
            }
        });
    }

    @ReactMethod
    public void online(final Promise promise) {
        executor.execute(new Runnable() {
            @Override
            public void run() {
                try {
                    promise.resolve(Textile.instance().online());
                }
                catch (final Exception e) {
                    promise.reject("online", e);
                }
            }
        });
    }

    @Override
    public void onCatalystInstanceDestroy() {
        super.onCatalystInstanceDestroy();
        executor.execute(new Runnable() {
            @Override
            public void run() {
                try {
                    Textile.instance().destroy();
                } catch (final Exception e) {
                    // noop
                }
            }
        });
    }
}
